### PR TITLE
helper functions for env_var emulation in Windows

### PR DESCRIPTION
**Pull request by JOE1994**
Moving some of the changes submitted in PR #1225, in order to prevent the original PR from bloating too much.

### CODE REVIEW

**Comment by RalfJung** [comment]
Thanks for splitting!

File: src/helpers.rs [hunk 1]
```diff
@@ -456,6 +456,17 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         }
     }
 
+    /// Dispatches to appropriate implementations for reading an OsString from Memory,
+    /// depending on the interpretation target.
+    fn read_os_str_from_target_str(&self, scalar: Scalar<Tag>) -> InterpResult<'tcx, OsString> {
+        let target_os = self.eval_context_ref().tcx.sess.target.target.target_os.as_str();
+        match target_os {
+            "linux" | "macos" => self.read_os_str_from_c_str(scalar).map(|x| x.to_os_string()),
+            "windows" => { self.read_os_str_from_wide_str(scalar) },
```
Comment (RalfJung) [inline src/helpers.rs]: Why the curly braces here?

File: src/helpers.rs [hunk 2]
```diff
@@ -456,6 +456,17 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         }
     }
 
+    /// Dispatches to appropriate implementations for reading an OsString from Memory,
+    /// depending on the interpretation target.
+    fn read_os_str_from_target_str(&self, scalar: Scalar<Tag>) -> InterpResult<'tcx, OsString> {
```
Comment (RalfJung) [inline src/helpers.rs]: If you want, this would be a nice opportunity to learn how to use `Cow`. You can make the type of this function
```rust
fn read_os_str_from_c_str<'a>(&'a self, scalar: Scalar<Tag>) -> InterpResult<'tcx, Cow<'a, OsStr>>
    where
        'tcx: 'a,
        'mir: 'a,
```
and then in the linux/macos case do `map(Cow::Borrowed)` and in the other case `.map(Cow::Owned)`. That way, we can avoid unnecessary copies.

File: src/helpers.rs [hunk 3]
```diff
@@ -518,21 +537,106 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
+#[cfg(target_os = "windows")]
+pub fn u16vec_to_osstring<'tcx>(u16_vec: Vec<u16>) -> InterpResult<'tcx, OsString> {
+    Ok(std::os::windows::ffi::OsStringExt::from_wide(&u16_vec[..]))
+}
+#[cfg(not(target_os = "windows"))]
+pub fn u16vec_to_osstring<'tcx>(u16_vec: Vec<u16>) -> InterpResult<'tcx, OsString> {
+    let s = String::from_utf16(&u16_vec[..])
+        .map_err(|_| err_unsup_format!("{:?} is not a valid utf-16 string", u16_vec))?;
+    Ok(s.into())
+}
```
Comment (RalfJung) [inline src/helpers.rs]: These functions are only used inside `read_os_str_from_wide_str` right? If yes, you cold move the helper functions inside that function, too, like you already did for `write_os_str_to_wide_str`.

File: src/helpers.rs [hunk 4]
```diff
@@ -518,21 +537,106 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
+    /// to write if `size` is not large enough to fit the contents of `os_string` plus a null
+    /// terminator. It returns `Ok((true, length))` if the writing process was successful. The
+    /// string length returned does not include the null terminator.
+    fn write_os_str_to_wide_str(
+        &mut self,
+        os_str: &OsStr,
+        mplace: MPlaceTy<'tcx, Tag>,
+        size: u64,
+    ) -> InterpResult<'tcx, (bool, u64)> {
+        #[cfg(target_os = "windows")]
```
Comment (RalfJung) [inline src/helpers.rs]: AFAIK one can also do `cfg(windows)`. Do you know what is preferred?

File: src/helpers.rs [hunk 5]
```diff
@@ -518,21 +537,106 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
+        fn os_str_to_u16vec<'tcx>(os_str: &OsStr) -> InterpResult<'tcx, Vec<u16>> {
+            Ok(std::os::windows::ffi::OsStrExt::encode_wide(os_str).collect())
+        }
+        #[cfg(not(target_os = "windows"))]
+        fn os_str_to_u16vec<'tcx>(os_str: &OsStr) -> InterpResult<'tcx, Vec<u16>> {
+            // On non-unix platforms the best we can do to transform Vec<u16> from/to OS strings is to do the
```
Comment (RalfJung) [inline src/helpers.rs]:
```suggestion
            // On non-Windows platforms the best we can do to transform Vec<u16> from/to OS strings is to do the
```

File: src/helpers.rs [hunk 6]
```diff
@@ -518,21 +537,106 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
+            os_str
+                .to_str()
+                .map(|s| s.encode_utf16().collect())
+                .ok_or_else(|| err_unsup_format!("{:?} is not a valid utf-8 string", os_str).into())
+        }
+
+        let u16_vec = os_str_to_u16vec(os_str)?;
+        // If `size` is smaller or equal than `bytes.len()`, writing `bytes` plus the required
+        // 0x0000 terminator to memory would cause an out-of-bounds access.
+        let string_length = u16_vec.len() as u64;
```
Comment (RalfJung) [inline src/helpers.rs]:
```suggestion
        let string_length = u64::try_from(u16_vec.len()).unwrap();
```

File: src/helpers.rs [hunk 7]
```diff
@@ -518,21 +537,106 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
-        arg_place.ptr.assert_ptr()
+        Ok(arg_place.ptr.assert_ptr())
+    }
+
+    fn alloc_os_str_as_wide_str(
+        &mut self,
+        os_str: &OsStr,
+        memkind: MemoryKind<MiriMemoryKind>,
+    ) -> InterpResult<'tcx, Pointer<Tag>> {
+        let size = os_str.len() as u64 + 1; // Make space for `0x0000` terminator.
```
Comment (RalfJung) [inline src/helpers.rs]:
```suggestion
        let size = u64::try_from(os_str.len()).unwrap().checked_add(1).unwrap(); // Make space for `0x0000` terminator.
```

File: src/shims/env.rs [hunk 8]
```diff
@@ -74,16 +74,16 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
 
         let name_ptr = this.read_scalar(name_op)?.not_undef()?;
         let value_ptr = this.read_scalar(value_op)?.not_undef()?;
-        let value = this.read_os_str_from_c_str(value_ptr)?;
+        let value = this.read_os_str_from_target_str(value_ptr)?;
```
Comment (RalfJung) [inline src/shims/env.rs]: If you add these here, could you also add the `target_os` assertion in `getenv`?

File: src/helpers.rs [hunk 9]
```diff
@@ -518,21 +537,106 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
         os_str: &OsStr,
         memkind: MemoryKind<MiriMemoryKind>,
-    ) -> Pointer<Tag> {
+    ) -> InterpResult<'tcx, Pointer<Tag>> {
         let size = u64::try_from(os_str.len()).unwrap().checked_add(1).unwrap(); // Make space for `0` terminator.
         let this = self.eval_context_mut();
 
         let arg_type = this.tcx.mk_array(this.tcx.types.u8, size);
         let arg_place = this.allocate(this.layout_of(arg_type).unwrap(), memkind);
         self.write_os_str_to_c_str(os_str, arg_place.ptr, size).unwrap();
```
Comment (RalfJung) [inline src/helpers.rs]:
```suggestion
        assert!(self.write_os_str_to_c_str(os_str, arg_place.ptr, size).unwrap().0);
```

File: src/helpers.rs [hunk 10]
```diff
@@ -518,21 +537,106 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
+        &mut self,
+        os_str: &OsStr,
+        memkind: MemoryKind<MiriMemoryKind>,
+    ) -> InterpResult<'tcx, Pointer<Tag>> {
+        let size = os_str.len() as u64 + 1; // Make space for `0x0000` terminator.
+        let this = self.eval_context_mut();
+
+        let arg_type = this.tcx.mk_array(this.tcx.types.u16, size);
+        let arg_place = this.allocate(this.layout_of(arg_type).unwrap(), memkind);
+        self.write_os_str_to_wide_str(os_str, arg_place, size).unwrap();
```
Comment (RalfJung) [inline src/helpers.rs]:
```suggestion
        assert!(self.write_os_str_to_wide_str(os_str, arg_place, size).unwrap().0);
```

File: src/helpers.rs [hunk 11]
```diff
@@ -518,21 +537,106 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
+            _ => throw_unsup_format!("OsString support for target OS not yet available"),
+        }
+    }
+
     fn alloc_os_str_as_c_str(
         &mut self,
         os_str: &OsStr,
         memkind: MemoryKind<MiriMemoryKind>,
-    ) -> Pointer<Tag> {
+    ) -> InterpResult<'tcx, Pointer<Tag>> {
```
Comment (RalfJung) [inline src/helpers.rs]: Why did you add `InterpResult` here? The method is not fallible.

File: src/helpers.rs [hunk 12]
```diff
@@ -518,21 +537,106 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
         self.write_os_str_to_c_str(os_str, arg_place.ptr, size).unwrap();
-        arg_place.ptr.assert_ptr()
+        Ok(arg_place.ptr.assert_ptr())
+    }
+
+    fn alloc_os_str_as_wide_str(
+        &mut self,
+        os_str: &OsStr,
+        memkind: MemoryKind<MiriMemoryKind>,
+    ) -> InterpResult<'tcx, Pointer<Tag>> {
```
Comment (RalfJung) [inline src/helpers.rs]: Why the `InterpResult` here?

File: src/helpers.rs (hunk 2 above)
Comment (RalfJung) [reply src/helpers.rs]: We can also do the Cow thing later though as you suggested. In that case, please leave a comment like `FIXME: Use Cow to avoid copies`.

File: src/helpers.rs (hunk 4 above)
Comment (JOE1994) [reply src/helpers.rs]: I'm not sure which is preferred, but `#[cfg(windows)]` looks more concise and desirable to me.

Comment (RalfJung) [reply src/helpers.rs]: That's also what I used in my WIP patch, so let's consistently use `cfg(windows)` then.

**Comment by JOE1994** [comment]
I force-pushed changes to resolve issues suggested in the reviews.

File: src/helpers.rs [hunk 13]
```diff
@@ -456,29 +456,64 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
+
     /// Helper function to read an OsString from a null-terminated sequence of bytes, which is what
     /// the Unix APIs usually handle.
     fn read_os_str_from_c_str<'a>(&'a self, scalar: Scalar<Tag>) -> InterpResult<'tcx, &'a OsStr>
     where
         'tcx: 'a,
         'mir: 'a,
     {
-        #[cfg(target_os = "unix")]
+        #[cfg(unix)]
```
Comment (RalfJung) [inline src/helpers.rs]: Something is weird with the diff view... this change already landed in master. could you rebase, maybe that fixes it?

Comment (JOE1994) [reply src/helpers.rs]: Thank you for mentioning the issue.. I made another force-push, and now the diff view looks okay to me

**Comment by RalfJung** [comment]
Looking good, thanks!

@bors r+

File: src/helpers.rs [hunk 14]
```diff
@@ -456,6 +456,18 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         }
     }
 
+    /// Dispatches to appropriate implementations for reading an OsString from Memory,
+    /// depending on the interpretation target.
+    /// FIXME: Use `Cow` to avoid copies
+    fn read_os_str_from_target_str(&self, scalar: Scalar<Tag>) -> InterpResult<'tcx, OsString> {
```
Comment (eddyb) [inline src/helpers.rs]: The `target_str` convention isn't great IMO, how about `target_encoding`? `wide_str`, be it UTF-16 or UCS-2, is an *encoding*, while WTF-8 is another.

Comment (RalfJung) [reply src/helpers.rs]: That's even longer though... `_from_target`?
